dash-pipeline/bmv2/stages/tunnel_stage.p4

```diff
@@ -0,0 +1,41 @@
+#ifndef _DASH_STAGE_TUNNEL_P4_
+#define _DASH_STAGE_TUNNEL_P4_
+
+control tunnel_stage(
+    inout headers_t hdr,
+    inout metadata_t meta)
+{
+    action set_tunnel_attrs(
+            @SaiVal[type="sai_ip_address_t"]
+            IPv4Address underlay_dip,
+            @SaiVal[type="sai_dash_encapsulation_t", default_value="SAI_DASH_ENCAPSULATION_VXLAN"]
+            dash_encapsulation_t dash_encapsulation,
+            bit<24> tunnel_key) {
+    push_action_static_encap(hdr = hdr,
```

Comment:
Hi Marian, after multiple rounds of thinking, I think it will be better to actually create a dedicated encap data in the metadata struct for tunnel instead of reusing the static encap one. Although the current approach works, but having a dedicated context will make the structure and logic cleaner.

Basically, the proposed update is:
- Create a new encap_data_t field for tunnel.
- Move the tunnel to the pre routing action stage instead of after action apply (which is confusing when looking at the pipeline).